src/Microsoft.ML.Tokenizers/Model/WordPieceTokenizer.cs

```diff
@@ -0,0 +1,858 @@
+                    IReadOnlyDictionary<string, int>? specialTokens,
+                    string unknownToken,
+                    string continuingSubwordPrefix = DefaultContinuingSubwordPrefix,
+                    int maxInputCharsPerWord = DefaultMaxInputCharsPerWord)
+        {
+            Debug.Assert(vocab is not null);
+            Debug.Assert(vocabReverse is not null);
+            _vocab = vocab!;
+            _vocabReverse = vocabReverse!;
+            SpecialTokens = specialTokens;
+            SpecialTokensReverse = specialTokens is not null ? specialTokens.ToDictionary(kvp => kvp.Value, kvp => kvp.Key) : null;
+
+            if (unknownToken is null)
+            {
+                throw new ArgumentNullException(nameof(unknownToken));
+            }
+
+            if (continuingSubwordPrefix is null)
+            {
+                throw new ArgumentNullException(nameof(continuingSubwordPrefix));
+            }
+
+            if (maxInputCharsPerWord <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxInputCharsPerWord), "The maximum number of characters per word must be greater than zero.");
+            }
+
+            if (!vocab!.TryGetValue(unknownToken, out int id))
+            {
+                throw new ArgumentException($"The unknown token '{unknownToken}' is not in the vocabulary.");
+            }
+
+            UnknownToken = unknownToken;
+            UnknownTokenId = id;
+            ContinuingSubwordPrefix = continuingSubwordPrefix;
+            MaxInputCharsPerWord = maxInputCharsPerWord;
+
+            _preTokenizer = preTokenizer ?? PreTokenizer.CreateWhiteSpacePreTokenizer(specialTokens);
+            _normalizer = normalizer;
+        }
+
+        /// <summary>
+        /// Gets the unknown token ID.
+        /// A token that is not in the vocabulary cannot be converted to an ID and is set to be this token instead.
+        /// </summary>
+        public int UnknownTokenId { get; }
+
+        /// <summary>
+        /// Gets the prefix to use for sub-words that are not the first part of a word.
+        /// </summary>
+        public string ContinuingSubwordPrefix { get; }
+
+        /// <summary>
+        /// Gets the maximum number of characters to authorize in a single word.
+        /// </summary>
+        public int MaxInputCharsPerWord { get; }
+
+        internal static async ValueTask<(Dictionary<StringSpanOrdinalKey, int>, Dictionary<int, string>)> LoadVocabAsync(Stream vocabStream, bool useAsync, CancellationToken cancellationToken = default)
+        {
+            if (vocabStream is null)
+            {
+                throw new ArgumentNullException(nameof(vocabStream));
+            }
+
+            Dictionary<StringSpanOrdinalKey, int> vocab = new Dictionary<StringSpanOrdinalKey, int>();
+            Dictionary<int, string> vocabReverse = new Dictionary<int, string>();
+
+            StreamReader reader = new StreamReader(vocabStream);
```

Comment:
This doesn't need to be disposed of? What are the ownership requirements of vocabStream?

Reply:
No, we cannot dispose the reader because it will dispose the stream which we need to avoid. The ownership and disposing the stream is done in the callers of the method (i.e. in Create method).

```
            try
            {
                (Dictionary<StringSpanOrdinalKey, int> vocab, Dictionary<int, string> vocabReverse) = await LoadVocabAsync(vocabStream, useAsync: true, cancellationToken);

                return new WordPieceTokenizer(vocab, vocabReverse, preTokenizer, normalizer, specialTokens, unknownToken, continuingSubwordPrefix, maxInputCharsPerWord);
            }
            finally
            {
                if (disposeStream)
                {
                    vocabStream.Dispose();
                }
            }

```